### DISCUSSION TITLE
Setting up next/router in react component library

### DISCUSSION THREAD

**Discussion by damiangene** [vercel/next.js#17554] [Help] [answered]
I'm trying to make use of Next/router in a React component library so that I can add routing logic into my React components. This library will only ever be consumed inside a Next.js project.

The component library isn't a Next.js project but I have added Next.js as a peer dependency.

When using any part of ```userRouter()``` I always get null errors.

```js
export const button = () => {
    const router = useRouter()
    
    const onClick = () => {
        console.log(router.pathName)
        router.push("/home")
    }
```

Both of these error out saying : 

```Cannot read property 'pathName' of null```

```Cannot read property 'push' of null```

I suspect this is because I can't use Next/router while not specifically running a Next project but is there a way around this, or am I missing something?

**Comment by jamesmosier** [comment, marked as answer]
> Next/router while not specifically running a Next project

Correct, `next/router` relies on being a child of a React Context Provider. 

> is there a way around this

Not that I know of. I believe all the various `next/` components must be apart of the same Next.js project. So I'm not certain how a component library with Next components in them would work 🤔

**Reply by damiangene** [reply]
Thank you for your response. That clears it all up thank you.

I managed to find a workaround. I added a property to my component that accepts a function with a parameter I then pass the router function into the component as a property. Seems to be working.

**Reply by Nick-Mazuk** [reply]
I ran into similar issues (if I understand the problem correctly), and I used both peer dependencies and dev dependencies.

By installing next + react as a dev dependency, everything resolves in the component library repo. Then by adding next + react as peer dependencies, the component library will use the project's versions of next and react.

As far as I know, dev dependencies aren't installed when doing `npm install [package]`, so this method shouldn't create bloat.
